### PR TITLE
Avoid copy for merge operator result using delete_callback

### PR DESCRIPTION
**Pull request by xuchen-plus**
This PR uses `delete_callback` to release memory of merge operator result.

The `rocksdb_mergeoperator_t` from Rocksdb C API allows a `delete_callback` argument which allows user to handle memory release [db/c.cc](https://github.com/facebook/rocksdb/blob/master/db/c.cc#L372)
The merge result `Vec<u8>` is turned into a boxed slice and turned into raw pointer by `Box::into_raw`. In `delete_callback`, `Box::from_raw` is called to wrap the raw pointer back and the original memory will be released here.

Testing done:
All unit tests passed.

Fixes: #410

### CODE REVIEW

File: src/merge_operator.rs [hunk 1]
```diff
@@ -101,15 +107,10 @@ pub unsafe extern "C" fn full_merge_callback(
...
-        result.shrink_to_fit();
-        // TODO(tan) investigate zero-copy techniques to improve performance
-        let buf = libc::malloc(result.len() as size_t);
-        assert!(!buf.is_null());
+    if let Some(result) = (cb.full_merge_fn)(key, oldval, operands) {
         *new_value_length = result.len() as size_t;
         *success = 1 as u8;
-        ptr::copy(result.as_ptr() as *mut c_void, &mut *buf, result.len());
-        buf as *mut c_char
+        (*Box::into_raw(result.into_boxed_slice())).as_ptr() as *mut c_char
```
Comment (aleksuss) [inline src/merge_operator.rs:113]:
```suggestion
        Box::into_raw(result.into_boxed_slice()) as *mut c_char
```

File: src/merge_operator.rs [hunk 2]
```diff
@@ -101,15 +107,10 @@ pub unsafe extern "C" fn full_merge_callback(
...
     } else {
         *new_value_length = 0;
         *success = 0 as u8;
```
Comment (aleksuss) [inline src/merge_operator.rs:116]:
```suggestion
        *success = 0u8;
```

File: src/merge_operator.rs [hunk 3]
```diff
@@ -130,15 +131,10 @@ pub unsafe extern "C" fn partial_merge_callback(
...
-        result.shrink_to_fit();
-        // TODO(tan) investigate zero-copy techniques to improve performance
-        let buf = libc::malloc(result.len() as size_t);
-        assert!(!buf.is_null());
+    if let Some(result) = (cb.partial_merge_fn)(key, None, operands) {
         *new_value_length = result.len() as size_t;
         *success = 1 as u8;
-        ptr::copy(result.as_ptr() as *mut c_void, &mut *buf, result.len());
-        buf as *mut c_char
+        (*Box::into_raw(result.into_boxed_slice())).as_ptr() as *mut c_char
```
Comment (aleksuss) [inline src/merge_operator.rs:137]:
```suggestion
        Box::into_raw(result.into_boxed_slice()) as *mut c_char
```

File: src/merge_operator.rs [hunk 4]
```diff
@@ -130,15 +131,10 @@ pub unsafe extern "C" fn partial_merge_callback(
...
     let operands = &mut MergeOperands::new(operands_list, operands_list_len, num_operands);
     let key = slice::from_raw_parts(raw_key as *const u8, key_len as usize);
-    if let Some(mut result) = (cb.partial_merge_fn)(key, None, operands) {
-        result.shrink_to_fit();
-        // TODO(tan) investigate zero-copy techniques to improve performance
-        let buf = libc::malloc(result.len() as size_t);
-        assert!(!buf.is_null());
+    if let Some(result) = (cb.partial_merge_fn)(key, None, operands) {
         *new_value_length = result.len() as size_t;
         *success = 1 as u8;
```
Comment (aleksuss) [inline src/merge_operator.rs:136]:
```suggestion
        *success = 1u8;
```

File: src/merge_operator.rs [hunk 5]
```diff
@@ -130,15 +131,10 @@ pub unsafe extern "C" fn partial_merge_callback(
...
-        ptr::copy(result.as_ptr() as *mut c_void, &mut *buf, result.len());
-        buf as *mut c_char
+        (*Box::into_raw(result.into_boxed_slice())).as_ptr() as *mut c_char
     } else {
         *new_value_length = 0;
         *success = 0 as u8;
```
Comment (aleksuss) [inline src/merge_operator.rs:140]:
```suggestion
        *success = 0u8;
```

**Review by aleksuss — COMMENTED** [review]
Overall looks good 👍🏻

**Review by aleksuss — APPROVED** [review]

**Review by stanislav-tkach — APPROVED** [review]

**Pull request merged**